README.md
Scrat.js
--------

## 介绍
Scrat.js 是与 [Scrat](https://github.com/scrat-team/scrat) 配套使用的前端模块管理框架，Scrat 通过 release 环节将计算出的模块列表、别名、依赖关系、配置等信息传递给 Scrat.js，由 Scrat.js 负责运行时的模块加载、合并请求、缓存等工作。

## 主要 API 说明
### require.async(modules, callback)
说明：加载并运行一组 JS 模块

- @param {string|string[]} modules - 要加载并运行的模块列表
- @param {Function} [callback] - 全部模块及其依赖加载成功后的回调函数

示例：
```javascript
require.async(['ajax', 'event'], function (ajax, event) {
    ajax.get('/someObjs', {length: 10}, function (data) {
        event.emit('done', data);
    });
});
```

### require.config(options)
说明：设置并返回 Scrat.js 选项

- @param {Object} [options] - 配置选项
- @returns {Object} options

示例：
```javascript
require.config(__FRAMEWORK_CONFIG__); // Scrat 在编译过程中会自动替换 __FRAMEWORK_CONFIG__ 为配置数据
require.config({
    cache: true, // 开启 localStorage 缓存
    urlPattern: '/path/to/resources/%s', // 资源加载路径
    comboPattern: '/path/to/combo-service??%s' // Combo 服务路径
});
```

### define(id, factory [, cache=true])
说明：定义一个 JS 模块

- @param {string} id - 模块 id
- @param {Function} factory - 模块的工厂函数
- @param {boolean} [cache=true] - 为 false 时候强制不缓存到 localStorage

示例：
```javascript
define('hello', function (require, exports, module) {
    module.exports = function (name) {
        alert('Hello ' + name);
    };
});
```

HISTORY.md
0.1.0 / 2014-03-05
------------------

* 完成首个可用版本开发

0.2.0 / 2014-03-16
------------------

* 重写大部分逻辑，优化模块加载排重逻辑，
* 新增 require.alias
* 新增 require.load
* 增加调试信息

0.3.0 / 2014-05-03
------------------

* 增加 localStorage 对模块 JS 文件的缓存
* 增加构建标识 hash，hash 发生变化时清空 localStorage 缓存

0.3.1 / 2014-05-07
------------------

* 增加 localStorage 对模块 CSS 文件的缓存，需与 scrat@0.1.12 配合使用

0.3.2 / 2014-05-18
------------------

* 增加对 combo-url 长度的限制，超长则对 combo 请求进行拆分

0.3.3 / 2014-05-23
------------------

* 增加 scrat.traceback，保存在执行模块 factory 过程中未捕获的异常
* 增加通过 GET 参数关闭 cache 或 combo 的能力，scrat=nocache,nocombo

0.3.4 / 2014-05-26
------------------

* 调整判断文件类型函数，原逻辑在 QQ 浏览器下无法正常拿到分组

0.3.5 / 2014-07-14
------------------

* 调整 link / scrpit 标签插入 head 的顺序

0.3.6 / 2014-10-24
------------------

* 调整 localStorage 中缓存的 JavaScript 代码运行方式，eval 会在 UC 浏览器 9.7.6 ~ 9.8.5 版本中无法调起浏览器自有功能

0.3.7 / 2014-11-07
------------------

* require.async 回调在 scrat 加载资源失败时也会执行，未加载到模块返回 null

0.3.8 / 2014-11-07
------------------

* 修正 res.onload 不存在导致报错问题
* node.onerror 增加堆栈输出

0.3.9 / 2015-04-12
------------------

* define 方法增加第三个参数，可以手动禁止模块被 localStorage 缓存

index.js
(function (global, undefined) {
    'use strict';

    var slice = Array.prototype.slice,
        proto = {},
        scrat = create(proto);

    scrat.options = {
        debug: false,
        timeout: 15, // seconds
        alias: {}, // key - name, value - id
        deps: {}, // key - id, value - name/id
        urlPattern: null, // '/path/to/resources/%s'
        comboPattern: null, // '/path/to/combo-service/%s' or function (ids) { return url; }
        combo: false
    };
    scrat.cache = {}; // key - id

    /**
     * Mix obj to scrat.options
     * @param {object} obj
     */
    proto.config = function (obj) {
        var options = scrat.options;

        debug('scrat.config', obj);
        each(obj, function (value, key) {
            var data = options[key],
                t = type(data);
            if (t === 'object') {
                each(value, function (v, k) {
                    data[k] = v;
                });
            } else {
                if (t === 'array') value = data.concat(value);
                options[key] = value;
            }
        });
    };

    /**
     * Require modules asynchronously with a callback
     * @param {string|array} names
     * @param {function} onload
     */
    proto.async = function (names, onload) {
        if (type(names) === 'string') names = [names];
        debug('scrat.async', 'require [' + names.join(', ') + ']');

        var reactor = new scrat.Reactor(names, function () {
            var args = [];
            each(names, function (id) {
                args.push(require(id));
            });
            onload.apply(scrat, args);
            debug('scrat.async', '[' + names.join(', ') + '] callback called');
        });
        reactor.run();
    };

    /**
     * Define a module with a factory funciton or any types of value
     * @param {string} id
     * @param {*} factory
     */
    proto.define = function (id, factory) {
        debug('scrat.define', '[' + id + ']');
        var res = scrat.cache[id];
        if (res) {
            res.factory = factory;
        } else {
            scrat.cache[id] = {
                id: id,
                loaded: true,
                factory: factory
            };
        }
    };

    /**
     * Get alias from specified name recursively
     * @param {string} name
     * @param {string|function} [alias] - set alias
     * @returns {string} name
     */
    proto.alias = function (name, alias) {
        var aliasMap = scrat.options.alias;

        if (arguments.length > 1) {
            aliasMap[name] = alias;
            return scrat.alias(name);
        }

        while (aliasMap[name] && name !== aliasMap[name]) {
            switch (type(aliasMap[name])) {
            case 'function':
                name = aliasMap[name](name);
                break;
            case 'string':
                name = aliasMap[name];
                break;
            }
        }
        return name;
    };

    /**
     * Load any types of resources from specified url
     * @param {string} url
     * @param {function|object} [onload|options]
     */
    proto.load = function (url, options) {
        if (type(options) === 'function') options = {onload: options};

        var t = options.type || fileType(url),
            isScript = t === 'js',
            isCss = t === 'css',
            isOldWebKit = +navigator.userAgent
                .replace(/.*AppleWebKit\/(\d+)\..*/, '$1') < 536,

            head = document.getElementsByTagName('head')[0],
            node = document.createElement(isScript ? 'script' : 'link'),
            supportOnload = 'onload' in node,
            tid = setTimeout(onerror, (options.timeout || 15) * 1000),
            intId, intTimer;

        if (isScript) {
            node.type = 'text/javascript';
            node.async = 'async';
            node.src = url;
        } else {
            if (isCss) {
                node.type = 'text/css';
                node.rel = 'stylesheet';
            }
            node.href = url;
        }

        node.onload = node.onreadystatechange = function () {
            if (!node.readyState ||
                /loaded|complete/.test(node.readyState)) {
                clearTimeout(tid);
                node.onload = node.onreadystatechange = null;
                if (isScript && head && node.parentNode) head.removeChild(node);
                if (options.onload) options.onload.call(scrat);
                node = null;
            }
        };

        node.onerror = function onerror() {
            clearTimeout(tid);
            clearInterval(intId);
            throw new Error('Error loading url: ' + url);
        };

        debug('scrat.load', '[' + url + ']');
        head.insertBefore(node, head.firstChild);

        // trigger onload immediately after nonscript node insertion
        if (isCss) {
            if (isOldWebKit || !supportOnload) {
                debug('scrat.load', 'check css\'s loading status for compatible');
                intTimer = 0;
                intId = setInterval(function () {
                    if ((intTimer += 20) > options.timeout || !node) {
                        clearTimeout(tid);
                        clearInterval(intId);
                        return;
                    }
                    if (node.sheet) {
                        clearTimeout(tid);
                        clearInterval(intId);
                        if (options.onload) options.onload.call(scrat);
                        node = null;
                    }
                }, 20);
            }
        } else if (!isScript) {
            if (options.onload) options.onload.call(scrat);
        }
    };

    proto.Reactor = function (names, callback) {
        this.length = 0;
        this.depends = {};
        this.depended = {};
        this.push.apply(this, names);
        this.callback = callback;
    };

    var rproto = scrat.Reactor.prototype;

    rproto.push = function () {
        var that = this,
            args = slice.call(arguments);

        function onload() {
            if (--that.length === 0) that.callback();
        }

        each(args, function (arg) {
            var id = scrat.alias(arg),
                type = fileType(id),
                res = scrat.cache[id];

            if (that.depended[id] || res && res.loaded) return;
            if (res && !res.loaded && type !== 'unknown') {
                ++that.length;
                res.onload.push(onload);
                return;
            }

            res = scrat.cache[id] = {
                id: id,
                loaded: false,
                onload: [onload]
            };

            that.push.apply(that, scrat.options.deps[id]);
            that.depended[id] = 1;
            if (!that.depends[type]) that.depends[type] = [];
            that.depends[type].push(res);
            if (type !== 'unknown') ++that.length;
        });
    };

    rproto.run = function () {
        var that = this,
            options = scrat.options,
            combo = options.combo,
            depends = this.depends;

        if (this.length === 0) return this.callback();
        debug('reactor.run', depends);

        each(depends.unknown, function (res) {
            scrat.load(that.genUrl(res.id), function () {
                res.loaded = true;
            });
        });

        debug('reactor.run', 'combo: ' + combo);
        if (combo) {
            each(['css', 'js'], function (type) {
                var ids = [];
                each(depends[type], function (res) {
                    ids.push(res.id);
                });
                if (ids.length) {
                    scrat.load(that.genUrl(ids), function () {
                        each(depends[type], function (res) {
                            res.loaded = true;
                            while (res.onload.length) {
                                var onload = res.onload.shift();
                                onload.call(res);
                            }
                        });
                    });
                }
            });
        } else {
            each((depends.css || []).concat(depends.js || []), function (res) {
                scrat.load(that.genUrl(res.id), function () {
                    res.loaded = true;
                    while (res.onload.length) {
                        var onload = res.onload.shift();
                        onload.call(res);
                    }
                });
            });
        }
    };

    rproto.genUrl = function (ids) {
        if (type(ids) === 'string') ids = [ids];

        var options = scrat.options,
            url = options.combo && options.comboPattern || options.urlPattern;
        switch (type(url)) {
        case 'string':
            url = url.replace('%s', ids.join(','));
            break;
        case 'function':
            url = url(ids);
            break;
        default:
            url = ids.join(',');
        }

        if (options.debug) {
            url = url + (~url.indexOf('?') ? '&' : '?') + (+new Date());
        }
        return url;
    };

    /**
     * Require another module in factory
     * @param {string} name
     * @returns {*} exports
     */
    function require(name) {
        var id = scrat.alias(name),
            module = scrat.cache[id];

        if (fileType(id) !== 'js') return;
        if (!module) throw new Error('failed to require "' + name + '"');

        if (!module.exports) {
            if (type(module.factory) === 'function') {
                module.factory.call(scrat, require, module.exports = {}, module);
            } else {
                module.exports = module.factory;
            }
            delete module.factory;
            debug('require', '[' + id + '] factory called');
        }
        return module.exports;
    }

    // Mix scrat's prototype to require
    each(proto, function (m, k) { require[k] = m; });

    function type(obj) {
        var t;
        if (obj == null) {
            t = String(obj);
        } else {
            t = Object.prototype.toString.call(obj).toLowerCase();
            t = t.substring(8, t.length - 1);
        }
        return t;
    }

    function each(obj, iterator, context) {
        if (typeof obj !== 'object') return;

        var i, l, t = type(obj);
        context = context || obj;
        if (t === 'array' || t === 'arguments' || t === 'nodelist') {
            for (i = 0, l = obj.length; i < l; i++) {
                if (iterator.call(context, obj[i], i, obj) === false) return;
            }
        } else {
            for (i in obj) {
                if (obj.hasOwnProperty(i)) {
                    if (iterator.call(context, obj[i], i, obj) === false) return;
                }
            }
        }
    }

    function create(proto) {
        function Dummy() {}
        Dummy.prototype = proto;
        return new Dummy();
    }

    var TYPE_RE = /(?:\.)(\w+)(?:[?&,]|$)/g;
    function fileType(str) {
        var ext = 'js',
            match = str.match(TYPE_RE);
        if (match && match.length) ext = RegExp.$1;
        if (ext === 'json') ext = 'js';
        else if (ext !== 'js' && ext !== 'css') ext = 'unknown';
        return ext;
    }

    var _modCache;
    function debug() {
        var flag = (global.localStorage || {}).debug,
            args = slice.call(arguments),
            style = 'color: #bada55',
            mod = args.shift(),
            re = new RegExp(mod.replace(/[.\/\\]/, function (m) {
                return '\\' + m;
            }));
        mod = '%c' + mod;
        if (flag && flag === '*' || re.test(flag)) {
            if (_modCache !== mod) {
                console.groupEnd(_modCache, style);
                console.group(_modCache = mod, style);
            }
            if (/string|number|boolean/.test(type(args[0]))) {
                args[0] = '%c' + args[0];
                args.splice(1, 0, style);
            }
            console.log.apply(console, args);
        }
    }

    global.require = scrat;
    global.define = scrat.define;
})(this);

scrat.js
(function (global) {
    'use strict';

    var slice = Array.prototype.slice,
        localStorage = global.localStorage,
        proto = {},
        scrat = create(proto);

    scrat.version = '0.3.9';
    scrat.options = {
        prefix: '__SCRAT__',
        cache: false,
        hash: '',
        timeout: 15, // seconds
        alias: {}, // key - name, value - id
        deps: {}, // key - id, value - name/id
        urlPattern: null, // '/path/to/resources/%s'
        comboPattern: null, // '/path/to/combo-service/%s' or function (ids) { return url; }
        combo: false,
        maxUrlLength: 2000 // approximate value of combo url's max length (recommend 2000)
    };
    scrat.cache = {}; // key - id
    scrat.traceback = null;

    /**
     * Mix obj to scrat.options
     * @param {Object} obj
     */
    proto.config = function (obj) {
        var options = scrat.options;

        debug('scrat.config', obj);
        each(obj, function (value, key) {
            var data = options[key],
                t = type(data);
            if (t === 'object') {
                each(value, function (v, k) {
                    data[k] = v;
                });
            } else {
                if (t === 'array') value = data.concat(value);
                options[key] = value;
            }
        });

        // detect localStorage support and activate cache ability
        try {
            if (options.hash !== localStorage.getItem('__SCRAT_HASH__')) {
                scrat.clean();
                localStorage.setItem('__SCRAT_HASH__', options.hash);
            }
            options.cache = options.cache && !!options.hash;
        } catch (e) {
            options.cache = false;
        }

        // detect scrat=nocombo,nocache in location.search
        if (/\bscrat=([\w,]+)\b/.test(location.search)) {
            each(RegExp.$1.split(','), function (o) {
                switch (o) {
                case 'nocache':
                    scrat.clean();
                    options.cache = false;
                    break;
                case 'nocombo':
                    options.combo = false;
                    break;
                }
            });
        }
        return options;
    };

    /**
     * Require modules asynchronously with a callback
     * @param {string|string[]} names
     * @param {Function} [onload]
     */
    proto.async = function (names, onload) {
        if (type(names) === 'string') names = [names];
        debug('scrat.async', 'require [' + names.join(', ') + ']');

        var reactor = new scrat.Reactor(names, function () {
            var args = [];
            each(names, function (id) {
                args.push(require(id));
            });
            if (onload) onload.apply(scrat, args);
            debug('scrat.async', '[' + names.join(', ') + '] callback called');
        });
        reactor.run();
    };

    /**
     * Define a JS module with a factory function
     * @param {string} id
     * @param {Function} factory
     * @param {boolean} [cache=true]
     */
    proto.define = function (id, factory, cache) {
        debug('scrat.define', '[' + id + ']');
        var options = scrat.options,
            res = scrat.cache[id];
        if (res) {
            res.factory = factory;
        } else {
            scrat.cache[id] = {
                id: id,
                loaded: true,
                factory: factory
            };
        }
        if (options.cache && cache !== false) {
            try {
                localStorage.setItem(options.prefix + id, factory.toString());
            } catch (e) {
                options.cache = false;
            }
        }
    };

    /**
     * Define a CSS module
     * @param {string} id
     * @param {string} css
     * @param {boolean} [parsing=true]
     */
    proto.defineCSS = function (id, css, parsing) {
        debug('scrat.defineCSS', '[' + id + ']');
        var options = scrat.options;
        scrat.cache[id] = {
            id: id,
            loaded: true,
            rawCSS: css
        };
        if (parsing !== false) requireCSS(id);
        if (options.cache) {
            try {
                localStorage.setItem(options.prefix + id, css);
            } catch (e) {
                options.cache = false;
            }
        }
    };

    /**
     * Get a defined module
     * @param {string} id
     * @returns {Object} module
     */
    proto.get = function (id) {
        /* jshint evil:true */
        debug('scrat.get', '[' + id + ']');
        var options = scrat.options,
            type = fileType(id),
            res = scrat.cache[id],
            raw;
        if (res) {
            return res;
        } else if (options.cache) {
            raw = localStorage.getItem(options.prefix + id);
            if (raw) {
                if (type === 'js') {
                    // Don't use eval or new Function in UC (9.7.6 ~ 9.8.5)
                    // global['eval'].call(global, 'define("' + id + '",' + raw + ')');
                    // new Function('define("' + id + '",' + raw + ')')();
                    var s = document.createElement('script');
                    s.appendChild(document.createTextNode('define("' + id + '",' + raw + ')'));
                    document.head.appendChild(s);
                } else if (type === 'css') {
                    scrat.defineCSS(id, raw, false);
                }
                scrat.cache[id].loaded = false;
                return scrat.cache[id];
            }
        }
        return null;
    };

    /**
     * Clean module cache in localStorage
     */
    proto.clean = function () {
        debug('scrat.clean');
        try {
            each(localStorage, function (_, key) {
                if (~key.indexOf(scrat.options.prefix)) {
                    localStorage.removeItem(key);
                }
            });
            localStorage.removeItem('__SCRAT_HASH__');
        } catch (e) {}
    };

    /**
     * Get alias from specified name recursively
     * @param {string} name
     * @param {string|Function} [alias] - set alias
     * @returns {string} name
     */
    proto.alias = function (name, alias) {
        var aliasMap = scrat.options.alias;

        if (arguments.length > 1) {
            aliasMap[name] = alias;
            return scrat.alias(name);
        }

        while (aliasMap[name] && name !== aliasMap[name]) {
            switch (type(aliasMap[name])) {
            case 'function':
                name = aliasMap[name](name);
                break;
            case 'string':
                name = aliasMap[name];
                break;
            }
        }
        return name;
    };

    /**
     * Load any types of resources from specified url
     * @param {string} url
     * @param {Function|Object} [onload|options]
     */
    proto.load = function (url, options) {
        if (type(options) === 'function') {
            options = {onload: options};
            if (type(arguments[2]) === 'function') options.onerror = arguments[2];
        }

        var t = options.type || fileType(url),
            isScript = t === 'js',
            isCss = t === 'css',
            isOldWebKit = +navigator.userAgent
                .replace(/.*AppleWebKit\/(\d+)\..*/, '$1') < 536,

            head = document.head,
            node = document.createElement(isScript ? 'script' : 'link'),
            supportOnload = 'onload' in node,
            tid = setTimeout(onerror, (options.timeout || 15) * 1000),
            intId, intTimer;

        if (isScript) {
            node.type = 'text/javascript';
            node.async = 'async';
            node.src = url;
        } else {
            if (isCss) {
                node.type = 'text/css';
                node.rel = 'stylesheet';
            }
            node.href = url;
        }

        node.onload = node.onreadystatechange = function () {
            if (node && (!node.readyState ||
                /loaded|complete/.test(node.readyState))) {
                clearTimeout(tid);
                node.onload = node.onreadystatechange = null;
                if (isScript && head && node.parentNode) head.removeChild(node);
                if (options.onload) options.onload.call(scrat);
                node = null;
            }
        };

        node.onerror = function onerror(e) {
            clearTimeout(tid);
            clearInterval(intId);
            e = (e || {}).error || new Error('load url timeout');
            e.message = 'Error loading url: ' + url + '. ' + e.message;
            if (options.onerror) options.onerror.call(scrat, e);
            else throw e;
        };

        debug('scrat.load', '[' + url + ']');
        head.appendChild(node);

        // trigger onload immediately after nonscript node insertion
        if (isCss) {
            if (isOldWebKit || !supportOnload) {
                debug('scrat.load', 'check css\'s loading status for compatible');
                intTimer = 0;
                intId = setInterval(function () {
                    if ((intTimer += 20) > options.timeout || !node) {
                        clearTimeout(tid);
                        clearInterval(intId);
                        return;
                    }
                    if (node.sheet) {
                        clearTimeout(tid);
                        clearInterval(intId);
                        if (options.onload) options.onload.call(scrat);
                        node = null;
                    }
                }, 20);
            }
        } else if (!isScript) {
            if (options.onload) options.onload.call(scrat);
        }
    };

    proto.Reactor = function (names, callback) {
        this.length = 0;
        this.depends = {};
        this.depended = {};
        this.push.apply(this, names);
        this.callback = callback;
    };

    var rproto = scrat.Reactor.prototype;

    rproto.push = function () {
        var that = this,
            args = slice.call(arguments);

        function onload() {
            if (--that.length === 0) that.callback();
        }

        each(args, function (arg) {
            var id = scrat.alias(arg),
                type = fileType(id),
                res = scrat.get(id);

            if (!res) {
                res = scrat.cache[id] = {
                    id: id,
                    loaded: false
                };
            } else if (that.depended[id] || res.loaded) return;
            if (!res.onload) res.onload = [];

            that.depended[id] = 1;
            that.push.apply(that, scrat.options.deps[id]);

            if ((type === 'css' && !res.rawCSS && !res.parsed) ||
                (type === 'js' && !res.factory && !res.exports)) {
                (that.depends[type] || (that.depends[type] = [])).push(res);
                ++that.length;
                res.onload.push(onload);
            } else if (res.rawCSS) {
                requireCSS(id);
            }
        });
    };

    function makeOnload(deps) {
        deps = deps.slice();
        return function (e) {
            if (e) error(e);
            each(deps, function (res) {
                if (!e) res.loaded = true;
                while (res.onload && res.onload.length) {
                    var onload = res.onload.shift();
                    onload.call(res);
                }
            });
        };
    }

    rproto.run = function () {
        var that = this,
            options = scrat.options,
            combo = options.combo,
            depends = this.depends;

        if (this.length === 0) return this.callback();
        debug('reactor.run', depends);

        each(depends.unknown, function (res) {
            scrat.load(that.genUrl(res.id), function () {
                res.loaded = true;
            });
        });

        debug('reactor.run', 'combo: ' + combo);
        if (combo) {
            each(['css', 'js'], function (type) {
                var urlLength = 0,
                    ids = [],
                    deps = [];

                each(depends[type], function (res, i) {
                    var onload;
                    if (urlLength + res.id.length < options.maxUrlLength) {
                        urlLength += res.id.length;
                        ids.push(res.id);
                        deps.push(res);
                    } else {
                        onload = makeOnload(deps);
                        scrat.load(that.genUrl(ids), onload, onload);
                        urlLength = res.id.length;
                        ids = [res.id];
                        deps = [res];
                    }
                    if (i === depends[type].length - 1) {
                        onload = makeOnload(deps);
                        scrat.load(that.genUrl(ids), onload, onload);
                    }
                });
            });
        } else {
            each((depends.css || []).concat(depends.js || []), function (res) {
                var onload = makeOnload([res]);
                scrat.load(that.genUrl(res.id), onload);
            });
        }
    };

    rproto.genUrl = function (ids) {
        if (type(ids) === 'string') ids = [ids];

        var options = scrat.options,
            url = options.combo && options.comboPattern || options.urlPattern;

        if (options.cache && fileType(ids[0]) === 'css') {
            each(ids, function (id, i) {
                ids[i] = id + '.js';
            });
        }

        switch (type(url)) {
        case 'string':
            url = url.replace('%s', ids.join(','));
            break;
        case 'function':
            url = url(ids);
            break;
        default:
            url = ids.join(',');
        }

        // The omission of `_hash=` might cause problem in wechat's webview
        return url + (~url.indexOf('?') ? '&' : '?') + '_hash=' + options.hash;
    };

    /**
     * Require another module in factory
     * @param {string} name
     * @returns {*} module.exports
     */
    function require(name) {
        var id = scrat.alias(name),
            module = scrat.get(id);

        if (fileType(id) !== 'js') return;
        if (!module) {
            error(new Error('failed to require "' + name + '"'));
            return null;
        }
        if (!module.exports) {
            if (type(module.factory) !== 'function') {
                throw new Error('failed to require "' + name + '"');
            }
            try {
                module.factory.call(scrat, require, module.exports = {}, module);
            } catch (e) {
                e.id = id;
                throw (scrat.traceback = e);
            }
            delete module.factory;
            debug('require', '[' + id + '] factory called');
        }
        return module.exports;
    }

    // Mix scrat's prototype to require
    each(proto, function (m, k) { require[k] = m; });

    /**
     * Parse CSS module
     * @param {string} name
     */
    function requireCSS(name) {
        var id = scrat.alias(name),
            module = scrat.get(id);

        if (fileType(id) !== 'css') return;
        if (!module) throw new Error('failed to require "' + name + '"');
        if (!module.parsed) {
            if (type(module.rawCSS) !== 'string') {
                throw new Error('failed to require "' + name + '"');
            }
            var styleEl = document.createElement('style');
            document.head.appendChild(styleEl);
            styleEl.appendChild(document.createTextNode(module.rawCSS));
            delete module.rawCSS;
            module.parsed = true;
        }
    }

    function type(obj) {
        var t;
        if (obj == null) {
            t = String(obj);
        } else {
            t = Object.prototype.toString.call(obj).toLowerCase();
            t = t.substring(8, t.length - 1);
        }
        return t;
    }

    function each(obj, iterator, context) {
        if (typeof obj !== 'object') return;

        var i, l, t = type(obj);
        context = context || obj;
        if (t === 'array' || t === 'arguments' || t === 'nodelist') {
            for (i = 0, l = obj.length; i < l; i++) {
                if (iterator.call(context, obj[i], i, obj) === false) return;
            }
        } else {
            for (i in obj) {
                if (obj.hasOwnProperty(i)) {
                    if (iterator.call(context, obj[i], i, obj) === false) return;
                }
            }
        }
    }

    function create(proto) {
        function Dummy() {}
        Dummy.prototype = proto;
        return new Dummy();
    }

    var TYPE_RE = /\.(js|css)(?=[?&,]|$)/i;
    function fileType(str) {
        var ext = 'js';
        str.replace(TYPE_RE, function (m, $1) {
            ext = $1;
        });
        if (ext !== 'js' && ext !== 'css') ext = 'unknown';
        return ext;
    }

    var _modCache;
    function debug() {
        var flag = (global.localStorage || {}).debug,
            args = slice.call(arguments),
            style = 'color: #bada55',
            mod = args.shift(),
            re = new RegExp(mod.replace(/[.\/\\]/g, function (m) {
                return '\\' + m;
            }));
        mod = '%c' + mod;
        if (flag && flag === '*' || re.test(flag)) {
            if (_modCache !== mod) {
                console.groupEnd(_modCache, style);
                console.group(_modCache = mod, style);
            }
            if (/string|number|boolean/.test(type(args[0]))) {
                args[0] = '%c' + args[0];
                args.splice(1, 0, style);
            }
            console.log.apply(console, args);
        }
    }

    function error() {
        if (console && type(console.error) === 'function') {
            console.error.apply(console, arguments);
        }
    }

    global.require = scrat;
    global.define = scrat.define;
    global.defineCSS = scrat.defineCSS;
    if (global.module && global.module.exports) global.module.exports = scrat;
})(this);